bridges/pangolin-pangolinparachain/bridge/src/service/subscribe.rs

```diff
@@ -0,0 +1,111 @@
+                tracing::info!(target: "pangolin-pangolinparachain", "Try to restart subscribtion service.");
+                execution = start().await;
+            }
+            Ok(())
+        });
+        Ok(Self { _greet })
+    }
+}
+
+async fn start() -> color_eyre::Result<()> {
+    let bridge_config: BridgeConfig = Config::restore(Names::BridgePangolinPangolinParachain)?;
+    let config_pangolin = bridge_config.pangolin;
+
+    let client_pangolin =
+        PangolinClientComponent::component(config_pangolin.to_pangolin_client_config()?).await?;
+    let client_rococo =
+        RococoClientComponent::component(bridge_config.rococo.to_rococo_client_config()?).await?;
+
+    let pangolin_handle = tokio::spawn(run_until_pangolin_connection_lost(client_pangolin));
+    let rococo_handle = tokio::spawn(run_until_rococo_connection_lost(client_rococo));
+    let (_result_p, _result_r) = (pangolin_handle.await, rococo_handle.await);
+    Ok(())
+}
+
+async fn run_until_pangolin_connection_lost(mut client: PangolinClient) -> color_eyre::Result<()> {
+    while let Err(err) = subscribe_pangolin(&client).await {
+        tracing::error!(target: "pangolin-pangolinparachain", "Failed to get get justification from pangolin: {:?}", err);
+        let bridge_config: BridgeConfig = Config::restore(Names::BridgePangolinPangolinParachain)?;
+        let client_pangolin =
+            PangolinClientComponent::component(bridge_config.pangolin.to_pangolin_client_config()?)
+                .await?;
+        client = client_pangolin;
+    }
+    Ok(())
+}
+
+async fn run_until_rococo_connection_lost(mut client: RococoClient) -> color_eyre::Result<()> {
+    while let Err(err) = subscribe_rococo(&client).await {
+        tracing::error!(target: "pangolin-pangolinparachain", "Failed to get justification from rococo: {:?}", err);
+        let bridge_config: BridgeConfig = Config::restore(Names::BridgePangolinPangolinParachain)?;
+        let client_rococo =
+            RococoClientComponent::component(bridge_config.rococo.to_rococo_client_config()?)
+                .await?;
+        client = client_rococo;
+    }
+    Ok(())
+}
+
+async fn subscribe_pangolin(client: &PangolinClient) -> color_eyre::Result<()> {
+    let mut subscribe = client.subscribe_grandpa_justifications().await?;
+    while let Some(justification) = subscribe.next().await {
+        let mut data = PANGOLIN_JUSTIFICATIONS.lock().await;
+        data.push_back(justification.unwrap());
+        if data.len() >= 100 {
+            data.pop_front();
+        }
+    }
+    Ok(())
+}
+
+async fn subscribe_rococo(client: &RococoClient) -> color_eyre::Result<()> {
+    let mut subscribe = client.subscribe_grandpa_justifications().await?;
+    while let Some(justification) = subscribe.next().await {
+        let mut data = ROCOCO_JUSTIFICATIONS.lock().await;
+        data.push_back(justification.unwrap());
+        if data.len() >= 100 {
```

Comment:
I think there should be pop_front first. because the capacity is 100 (`VecDeque::with_capacity(100)`),  If you push_back first, it will exceed 100.